### PR TITLE
Added functionality to allow user to add a message to a topic partition manually

### PR DESCRIPTION
**Pull request by nguyen-tri-nhan**
Resolve conflict for #230 

image
image
image

### CODE REVIEW

File: src/main/java/kafdrop/service/KafkaHighLevelConsumer.java [hunk 1]
```diff
@@ -204,11 +204,19 @@ synchronized List<ConsumerRecord<String, String>> getLatestRecords(String topic,
...
                                            Deserializers deserializers) {
     initializeClient();
-    final List<TopicPartition> partitions = determinePartitionsForTopic(topic);
+    List<TopicPartition> partitions = determinePartitionsForTopic(topic);
+    if (partition != -1) {
+      var partitionOpt = partitions.stream().filter(p -> p.partition() == partition).findAny();
+      if (partitionOpt.isEmpty()) {
+        throw new IllegalArgumentException("Partition does not exist in topic");
+      }
+      partitions = List.of(partitionOpt.get());
```
Comment (Bert-R) [inline src/main/java/kafdrop/service/KafkaHighLevelConsumer.java]: Make this a little more compact:
```suggestion
      partitions = List.of(partitionOpt.orElseThrow(() -> new IllegalArgumentException("Partition " + partition + " does not exist in topic")));
 ```

File: src/main/java/kafdrop/service/KafkaHighLevelProducer.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			properties.put(ProducerConfig.RETRIES_CONFIG, 0);
+			properties.put(ProducerConfig.LINGER_MS_CONFIG, 1);
+			properties.put(ProducerConfig.CLIENT_ID_CONFIG, "kafdrop-producer");
+			kafkaConfiguration.applyCommon(properties);
+			
+			kafkaProducer = new KafkaProducer<>(properties);
+		}
+	}
+	
+	public RecordMetadata publishMessage(CreateMessageVO message, Serializers searilazers) {
```
Comment (Bert-R) [inline src/main/java/kafdrop/service/KafkaHighLevelProducer.java]: Spelling error:
```suggestion
	public RecordMetadata publishMessage(CreateMessageVO message, Serializers serializers) {

```

File: src/main/java/kafdrop/service/KafkaMonitor.java [hunk 3]
```diff
@@ -28,6 +28,10 @@
 import kafdrop.model.TopicVO;
 import kafdrop.util.Deserializers;
 import org.apache.kafka.common.TopicPartition;
+import kafdrop.model.*;
+import kafdrop.util.*;
```
Comment (Bert-R) [inline src/main/java/kafdrop/service/KafkaMonitor.java]: Don't use wildcard imports as that is not our coding convention.

File: src/main/java/kafdrop/service/KafkaMonitorImpl.java [hunk 4]
```diff
@@ -18,41 +18,21 @@
...
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.*;
+import org.apache.kafka.common.header.*;
+import org.slf4j.*;
+import org.springframework.stereotype.*;
+
+import java.util.*;
+import java.util.Map.*;
+import java.util.function.*;
+import java.util.stream.*;
```
Comment (Bert-R) [inline src/main/java/kafdrop/service/KafkaMonitorImpl.java]: Don't use wildcard imports as that is not our coding convention.

File: src/main/resources/application.yml [hunk 5]
```diff
@@ -38,6 +38,7 @@ kafdrop.monitor:
 
 kafka:
   brokerConnect: localhost:9092
+  isSecured: false
```
Comment (Bert-R) [inline src/main/resources/application.yml]: Why is this added?

**Review by Bert-R — CHANGES_REQUESTED** [review]
@nguyen-tri-nhan Thanks for this contribution!

I've reviewed it and added a couple of small comments and a question

Besides these, our linter isn't happy, see the below log.

If you resolve these, I would be happy to merge this PR.

```
2024-12-05 19:29:28 [INFO]   Linting EDITORCONFIG items...
  Error: -05 19:29:29 [ERROR]   Found errors when linting EDITORCONFIG. Exit code: 1.
  2024-12-05 19:29:29 [INFO]   Command output for EDITORCONFIG:
  ------
  src/main/java/kafdrop/util/DefaultMessageSerializer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	5: Wrong indent style found (tabs instead of spaces)
  	6: Wrong indent style found (tabs instead of spaces)
  	7: Wrong indent style found (tabs instead of spaces)
  	8: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/controller/MessageController.java:
  	211: Wrong indent style found (tabs instead of spaces)
  	212: Wrong indent style found (tabs instead of spaces)
  	213: Wrong indent style found (tabs instead of spaces)
  	214: Wrong indent style found (tabs instead of spaces)
  	215: Wrong indent style found (tabs instead of spaces)
  	216: Wrong indent style found (tabs instead of spaces)
  	217: Wrong indent style found (tabs instead of spaces)
  	218: Wrong indent style found (tabs instead of spaces)
  	219: Wrong indent style found (tabs instead of spaces)
  	[220](https://github.com/obsidiandynamics/kafdrop/actions/runs/12143699503/job/33995776344?pr=703#step:4:222): Wrong indent style found (tabs instead of spaces)
  	221: Wrong indent style found (tabs instead of spaces)
  	222: Wrong indent style found (tabs instead of spaces)
  	[223](https://github.com/obsidiandynamics/kafdrop/actions/runs/12143699503/job/33995776344?pr=703#step:4:225): Wrong indent style found (tabs instead of spaces)
  	224: Wrong indent style found (tabs instead of spaces)
  	[225](https://github.com/obsidiandynamics/kafdrop/actions/runs/12143699503/job/33995776344?pr=703#step:4:227): Wrong indent style found (tabs instead of spaces)
  	227: Wrong indent style found (tabs instead of spaces)
  	228: Wrong indent style found (tabs instead of spaces)
  	229: Wrong indent style found (tabs instead of spaces)
  	230: Wrong indent style found (tabs instead of spaces)
  	231: Wrong indent style found (tabs instead of spaces)
  	233: Wrong indent style found (tabs instead of spaces)
  	234: Wrong indent style found (tabs instead of spaces)
  	235: Wrong indent style found (tabs instead of spaces)
  	236: Wrong indent style found (tabs instead of spaces)
  	237: Wrong indent style found (tabs instead of spaces)
  	238: Wrong indent style found (tabs instead of spaces)
  	240: Wrong indent style found (tabs instead of spaces)
  	241: Wrong indent style found (tabs instead of spaces)
  	242: Wrong indent style found (tabs instead of spaces)
  	243: Wrong indent style found (tabs instead of spaces)
  	244: Wrong indent style found (tabs instead of spaces)
  	245: Wrong indent style found (tabs instead of spaces)
  	253: Wrong indent style found (tabs instead of spaces)
  	254: Wrong indent style found (tabs instead of spaces)
  	255: Wrong indent style found (tabs instead of spaces)
  	256: Wrong indent style found (tabs instead of spaces)
  	440: Wrong indent style found (tabs instead of spaces)
  	441: Wrong indent style found (tabs instead of spaces)
  	442: Wrong indent style found (tabs instead of spaces)
  	443: Wrong indent style found (tabs instead of spaces)
  	444: Wrong indent style found (tabs instead of spaces)
  	446: Wrong indent style found (tabs instead of spaces)
  	447: Wrong indent style found (tabs instead of spaces)
  	448: Wrong indent style found (tabs instead of spaces)
  	449: Wrong indent style found (tabs instead of spaces)
  	450: Wrong indent style found (tabs instead of spaces)
  	451: Wrong indent style found (tabs instead of spaces)
  	452: Wrong indent style found (tabs instead of spaces)
  	453: Wrong indent style found (tabs instead of spaces)
  	454: Wrong indent style found (tabs instead of spaces)
  	455: Wrong indent style found (tabs instead of spaces)
  	456: Wrong indent style found (tabs instead of spaces)
  	457: Wrong indent style found (tabs instead of spaces)
  	458: Wrong indent style found (tabs instead of spaces)
  	459: Wrong indent style found (tabs instead of spaces)
  	460: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/MessageSerializer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	5: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/MsgPackMessageSerializer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	12: Wrong indent style found (tabs instead of spaces)
  	13: Wrong indent style found (tabs instead of spaces)
  	14: Wrong indent style found (tabs instead of spaces)
  	15: Wrong indent style found (tabs instead of spaces)
  	16: Wrong indent style found (tabs instead of spaces)
  	17: Wrong indent style found (tabs instead of spaces)
  	18: Wrong indent style found (tabs instead of spaces)
  	19: Wrong indent style found (tabs instead of spaces)
  	20: Wrong indent style found (tabs instead of spaces)
  	21: Wrong indent style found (tabs instead of spaces)
  	22: Wrong indent style found (tabs instead of spaces)
  	23: Wrong indent style found (tabs instead of spaces)
  	24: Wrong indent style found (tabs instead of spaces)
  	25: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/model/CreateMessageVO.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	9: Wrong indent style found (tabs instead of spaces)
  	10: Wrong indent style found (tabs instead of spaces)
  	11: Wrong indent style found (tabs instead of spaces)
  	12: Wrong indent style found (tabs instead of spaces)
  	13: Wrong indent style found (tabs instead of spaces)
  	14: Wrong indent style found (tabs instead of spaces)
  	15: Wrong indent style found (tabs instead of spaces)
  	16: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/ProtobufMessageSerializer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	24: Wrong indent style found (tabs instead of spaces)
  	25: Wrong indent style found (tabs instead of spaces)
  	27: Wrong indent style found (tabs instead of spaces)
  	29: Wrong indent style found (tabs instead of spaces)
  	30: Wrong indent style found (tabs instead of spaces)
  	31: Wrong indent style found (tabs instead of spaces)
  	32: Wrong indent style found (tabs instead of spaces)
  	34: Wrong indent style found (tabs instead of spaces)
  	35: Wrong indent style found (tabs instead of spaces)
  	36: Wrong indent style found (tabs instead of spaces)
  	37: Wrong indent style found (tabs instead of spaces)
  	39: Wrong indent style found (tabs instead of spaces)
  	40: Wrong indent style found (tabs instead of spaces)
  	41: Wrong indent style found (tabs instead of spaces)
  	42: Wrong indent style found (tabs instead of spaces)
  	43: Wrong indent style found (tabs instead of spaces)
  	44: Wrong indent style found (tabs instead of spaces)
  	46: Wrong indent style found (tabs instead of spaces)
  	47: Wrong indent style found (tabs instead of spaces)
  	49: Wrong indent style found (tabs instead of spaces)
  	50: Wrong indent style found (tabs instead of spaces)
  	51: Wrong indent style found (tabs instead of spaces)
  	52: Wrong indent style found (tabs instead of spaces)
  	53: Wrong indent style found (tabs instead of spaces)
  	54: Wrong indent style found (tabs instead of spaces)
  	55: Wrong indent style found (tabs instead of spaces)
  	56: Wrong indent style found (tabs instead of spaces)
  	57: Wrong indent style found (tabs instead of spaces)
  	58: Wrong indent style found (tabs instead of spaces)
  	59: Wrong indent style found (tabs instead of spaces)
  	60: Wrong indent style found (tabs instead of spaces)
  	61: Wrong indent style found (tabs instead of spaces)
  	62: Wrong indent style found (tabs instead of spaces)
  	63: Wrong indent style found (tabs instead of spaces)
  	64: Wrong indent style found (tabs instead of spaces)
  	65: Wrong indent style found (tabs instead of spaces)
  	66: Wrong indent style found (tabs instead of spaces)
  	67: Wrong indent style found (tabs instead of spaces)
  	68: Wrong indent style found (tabs instead of spaces)
  	69: Wrong indent style found (tabs instead of spaces)
  	70: Wrong indent style found (tabs instead of spaces)
  	71: Wrong indent style found (tabs instead of spaces)
  	72: Wrong indent style found (tabs instead of spaces)
  	73: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/SerializationException.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	5: Wrong indent style found (tabs instead of spaces)
  	6: Wrong indent style found (tabs instead of spaces)
  	7: Wrong indent style found (tabs instead of spaces)
  	8: Wrong indent style found (tabs instead of spaces)
  	9: Wrong indent style found (tabs instead of spaces)
  	10: Wrong indent style found (tabs instead of spaces)
  	11: Wrong indent style found (tabs instead of spaces)
  	12: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/service/KafkaHighLevelProducer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	24: Wrong indent style found (tabs instead of spaces)
  	26: Wrong indent style found (tabs instead of spaces)
  	28: Wrong indent style found (tabs instead of spaces)
  	30: Wrong indent style found (tabs instead of spaces)
  	31: Wrong indent style found (tabs instead of spaces)
  	32: Wrong indent style found (tabs instead of spaces)
  	34: Wrong indent style found (tabs instead of spaces)
  	35: Wrong indent style found (tabs instead of spaces)
  	36: Wrong indent style found (tabs instead of spaces)
  	37: Wrong indent style found (tabs instead of spaces)
  	38: Wrong indent style found (tabs instead of spaces)
  	39: Wrong indent style found (tabs instead of spaces)
  	40: Wrong indent style found (tabs instead of spaces)
  	41: Wrong indent style found (tabs instead of spaces)
  	42: Wrong indent style found (tabs instead of spaces)
  	43: Wrong indent style found (tabs instead of spaces)
  	44: Wrong indent style found (tabs instead of spaces)
  	45: Wrong indent style found (tabs instead of spaces)
  	46: Wrong indent style found (tabs instead of spaces)
  	47: Wrong indent style found (tabs instead of spaces)
  	48: Wrong indent style found (tabs instead of spaces)
  	49: Wrong indent style found (tabs instead of spaces)
  	50: Wrong indent style found (tabs instead of spaces)
  	51: Wrong indent style found (tabs instead of spaces)
  	52: Wrong indent style found (tabs instead of spaces)
  	53: Wrong indent style found (tabs instead of spaces)
  	54: Wrong indent style found (tabs instead of spaces)
  	55: Wrong indent style found (tabs instead of spaces)
  	56: Wrong indent style found (tabs instead of spaces)
  	58: Wrong indent style found (tabs instead of spaces)
  	59: Wrong indent style found (tabs instead of spaces)
  	60: Wrong indent style found (tabs instead of spaces)
  	61: Wrong indent style found (tabs instead of spaces)
  	62: Wrong indent style found (tabs instead of spaces)
  	63: Wrong indent style found (tabs instead of spaces)
  	64: Wrong indent style found (tabs instead of spaces)
  	65: Wrong indent style found (tabs instead of spaces)
  	66: Wrong indent style found (tabs instead of spaces)
  	67: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/Serializers.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	5: Wrong indent style found (tabs instead of spaces)
  	6: Wrong indent style found (tabs instead of spaces)
  	7: Wrong indent style found (tabs instead of spaces)
  	8: Wrong indent style found (tabs instead of spaces)
  	9: Wrong indent style found (tabs instead of spaces)
  	10: Wrong indent style found (tabs instead of spaces)
  	11: Wrong indent style found (tabs instead of spaces)
  	13: Wrong indent style found (tabs instead of spaces)
  	14: Wrong indent style found (tabs instead of spaces)
  	15: Wrong indent style found (tabs instead of spaces)
  	17: Wrong indent style found (tabs instead of spaces)
  	18: Wrong indent style found (tabs instead of spaces)
  	19: Wrong indent style found (tabs instead of spaces)
  	20: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/service/KafkaMonitorImpl.java:
  	49: Line too long (159 instead of 120)
  	406: Wrong indent style found (tabs instead of spaces)
  	407: Wrong indent style found (tabs instead of spaces)
  	408: Wrong indent style found (tabs instead of spaces)
  	409: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/service/KafkaProducerException.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	8: Wrong indent style found (tabs instead of spaces)
  	9: Wrong indent style found (tabs instead of spaces)
  	10: Wrong indent style found (tabs instead of spaces)
  	11: Wrong indent style found (tabs instead of spaces)
  	12: Wrong indent style found (tabs instead of spaces)
  	13: Wrong indent style found (tabs instead of spaces)
  	14: Wrong indent style found (tabs instead of spaces)
  	15: Wrong indent style found (tabs instead of spaces)
  src/main/java/kafdrop/util/AvroMessageSerializer.java:
  	Wrong line endings or no final newline
  	Not all lines have the correct end of line character
  	10: Wrong indent style found (tabs instead of spaces)
  	11: Wrong indent style found (tabs instead of spaces)
  	13: Wrong indent style found (tabs instead of spaces)
  	14: Wrong indent style found (tabs instead of spaces)
  	15: Wrong indent style found (tabs instead of spaces)
  	16: Wrong indent style found (tabs instead of spaces)
  	18: Wrong indent style found (tabs instead of spaces)
  	19: Wrong indent style found (tabs instead of spaces)
  	20: Wrong indent style found (tabs instead of spaces)
  	21: Wrong indent style found (tabs instead of spaces)
  	22: Wrong indent style found (tabs instead of spaces)
  	24: Wrong indent style found (tabs instead of spaces)
  	25: Wrong indent style found (tabs instead of spaces)
  	26: Wrong indent style found (tabs instead of spaces)
  	27: Wrong indent style found (tabs instead of spaces)
  	28: Wrong indent style found (tabs instead of spaces)
  	29: Wrong indent style found (tabs instead of spaces)
  	30: Wrong indent style found (tabs instead of spaces)
  	31: Wrong indent style found (tabs instead of spaces)
  	32: Wrong indent style found (tabs instead of spaces)
  	33: Wrong indent style found (tabs instead of spaces)
  	34: Wrong indent style found (tabs instead of spaces)
  
243 errors found
```

File: src/main/java/kafdrop/service/KafkaMonitor.java (hunk 3 above)
Comment (nguyen-tri-nhan) [reply src/main/java/kafdrop/service/KafkaMonitor.java]: I'm resolved it

File: src/main/java/kafdrop/service/KafkaMonitorImpl.java (hunk 4 above)
Comment (nguyen-tri-nhan) [reply src/main/java/kafdrop/service/KafkaMonitorImpl.java]: I'm resolved it

File: src/main/resources/application.yml (hunk 5 above)
Comment (nguyen-tri-nhan) [reply src/main/resources/application.yml]: I've checked it and removed as a redundant

File: src/main/java/kafdrop/service/KafkaHighLevelConsumer.java (hunk 1 above)
Comment (nguyen-tri-nhan) [reply src/main/java/kafdrop/service/KafkaHighLevelConsumer.java]: Thank you sir

File: src/main/java/kafdrop/service/KafkaHighLevelProducer.java (hunk 2 above)
Comment (nguyen-tri-nhan) [reply src/main/java/kafdrop/service/KafkaHighLevelProducer.java]: I've fixed it

**Comment by nguyen-tri-nhan** [comment]
Thank you for your review @Bert-R, I've resolved your request change and linter fail as well. Hope the linter feels happy

**Comment by nguyen-tri-nhan** [comment]
I am changing the IDE line separator for each file to LF, hope it run well with linter

**Review by Bert-R — APPROVED** [review]
@nguyen-tri-nhan Looks good. Thanks!

**Comment by ydan002** [comment]
@nguyen-tri-nhan very useful feature!, however, is there any way to have this as a feature flag so it can be disabled in production environment?

**Comment by Bert-R** [comment]
@ydan002 Feel free to create a PR to propose such a feature flag
